File: 3918.sum-of-primes-between-number-and-its-reverse/solution.go
func sumOfPrimesInRange(n int) int {
    r := reverse(n)
    p := min(n, r)
    q := max(n, r)
    
    sieve := make([]bool, q+1)
    sieve[1] = true

    for i := 4; i <= q; i += 2 {
        sieve[i] = true
    }
    for i := 3; i <= q; i++ {
        for j := 2*i; j <= q; j += i {
            sieve[j] = true
        }
    }

    sum := 0

    for i := p; i <= q; i++ {
        if !sieve[i] {
            sum += i
        }
    }

    return sum
}

func reverse(n int) int {
    r := 0
    for ; n > 0; n /= 10 {
        r = 10*r + n%10
    }
    return r
}

File: 3918.sum-of-primes-between-number-and-its-reverse/solution.kt
class Solution {
    fun sumOfPrimesInRange(n: Int): Int {
        val r = reverse(n)
        val p = minOf(n, r)
        val q = maxOf(n, r)

        return (p..q).filter { isPrime(it) }.sum()
    }

    fun reverse(n: Int): Int {
        var r = 0
        var i = n

        while (i > 0) {
            r = 10*r + i%10
            i /= 10
        }
        return r
    }

    fun isPrime(n: Int): Boolean {
        return when {
            n == 1 -> false
            n == 2 -> true
            n%2 == 0 -> false
            else -> isPrimeOdd3Plus(n)
        }
    }

    fun isPrimeOdd3Plus(n: Int): Boolean {
        var i = 3

        while (i*i <= n) {
            if (n%i == 0) {
                return false
            }
            i += 2
        }
        return true
    }
}